subprojects/core/src/main/java/org/gradle/internal/build/PlannedNodeGraph.java

```diff
@@ -0,0 +1,235 @@
+ * limitations under the License.
+ */
+
+package org.gradle.internal.build;
+
+import com.google.common.collect.ImmutableList;
+import org.gradle.execution.plan.Node;
+import org.gradle.execution.plan.PlannedNodeInternal;
+import org.gradle.execution.plan.ToPlannedNodeConverter;
+import org.gradle.execution.plan.ToPlannedNodeConverterRegistry;
+import org.gradle.internal.taskgraph.CalculateTaskGraphBuildOperationType.PlannedNode;
+import org.gradle.internal.taskgraph.CalculateTaskGraphBuildOperationType.TaskIdentity;
+import org.gradle.internal.taskgraph.NodeIdentity;
+import org.gradle.internal.taskgraph.NodeIdentity.NodeType;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.EnumSet;
+import java.util.HashMap;
+import java.util.IdentityHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Queue;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+import static java.util.Collections.newSetFromMap;
+
+/**
+ * A graph of planned nodes that supports extracting sub-graphs of different {@link DetailLevel levels of detail}
+ * depending on what node types are supported by the consumer.
+ *
+ * @see Collector
+ */
+public class PlannedNodeGraph {
+
+    private final DetailLevel detailLevel;
+    private final List<PlannedNodeInternal> plannedNodes;
+
+    public PlannedNodeGraph(DetailLevel detailLevel, List<PlannedNodeInternal> plannedNodes) {
+        this.detailLevel = detailLevel;
+        this.plannedNodes = ImmutableList.copyOf(plannedNodes);
+    }
+
+    /**
+     * Returns the nodes from a subgraph that are of the given or lower detail level
+     * (lower if the requested detail level is higher than the available detail level).
+     */
+    public List<? extends PlannedNode> getNodes(DetailLevel detailLevel) {
+        if (detailLevel.level >= this.detailLevel.level) {
+            return plannedNodes;
+        }
+
+        return computePlan(detailLevel);
+    }
+
+    private List<PlannedNodeInternal> computePlan(DetailLevel detailLevel) {
+        HashMap<NodeIdentity, List<? extends NodeIdentity>> plannedNodeByIdentity = new HashMap<>();
+        for (PlannedNodeInternal plannedNode : plannedNodes) {
+            plannedNodeByIdentity.put(plannedNode.getNodeIdentity(), plannedNode.getNodeDependencies());
+        }
+
+        Function<NodeIdentity, NodeIdentity> maybeIdentityProvider = id -> detailLevel.contains(id.getNodeType()) ? id : null;
+        Function<NodeIdentity, List<? extends NodeIdentity>> traverser = id -> {
+            List<? extends NodeIdentity> deps = plannedNodeByIdentity.get(id);
+            if (deps == null) {
+                throw new IllegalStateException("No dependencies for node: " + id);
+            }
+            return deps;
+        };
+
+        List<PlannedNodeInternal> newPlannedNodes = new ArrayList<>();
```

Comment:
💭 I suppose it isn't worth trying to allocate the right size for this array list since that would mean traversing the planned nodes twice, right?

Reply:
I guess we could allocate the right size, given we already traverse `plannedNodes` at the beginning of the method. However, I don't think it will make any measurable difference in performance: this method will be called effectively once per build and the algorithm itself relatively heavy.